### PR TITLE
HIVE-25207. Expose incremental load statistics via JMX.

### PR DESCRIPTION
**Pull request by ayushtkn**
JMX output in UI- First:
image
JMX output in UI- Second:
image
JMX output in UI- Third:
image

Sample Output from JConsole:
image

JConsole-MBeans Values:
image

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWorkMBean.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+   * Gets the stage of Replication.
+   * @return INCREMENTAL or BOOTSTRAP load.
+   */
+  public String getStage();
+
+  /**
+   * Gets the name of scheduled query being run.
+   * @return the name of the scheduled query.
+   */
+  public String getScheduledQueryName();
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWorkMBean.java:44]: Add current event id processed and last repl id?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWorkMBean.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.exec.repl;
+
+public interface ReplLoadWorkMBean {
```
[the remark this answers is not in the document]
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWorkMBean.java:20]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplStatsTracker.java [hunk 3]
```diff
@@ -29,6 +32,7 @@
  * Tracks the replication statistics per event type.
  */
 public class ReplStatsTracker {
+  private static final Logger LOG = LoggerFactory.getLogger(ReplStatsTracker.class);
```
[the remark this answers is not in the document]
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplStatsTracker.java]: Removed

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java [hunk 4]
```diff
@@ -4663,6 +4678,26 @@ public void testIncrementalStatisticsMetrics() throws Throwable {
...
+    MBeans.unregister(ObjectName.getInstance(nameStri));
+  }
+
+  private void verifyMBeanStatistics(String testName, String replDbName, String nameStri, String[] events)
+      throws MalformedObjectNameException, MBeanException, AttributeNotFoundException, InstanceNotFoundException,
+      ReflectionException {
+    ObjectName name = ObjectName.getInstance(nameStri);
+
+    assertTrue(
+        ManagementFactory.getPlatformMBeanServer().getAttribute(name, "Stage").toString().contains("INCREMENTAL"));
```
[the remark this answers is not in the document]
Comment (ayushtkn) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWorkMBean.java (hunk 1 above)
Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWorkMBean.java:44]: The repl stats tracks what comes to replLogger, the Last event id is there, I have added that in the JMX

**Comment by ayushtkn** [comment]
New Output with the added Fields:
**1st Snap:**
image

**2nd Snap:**
image

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWorkMBean.java (hunk 1 above)
Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWorkMBean.java:44]: what about current event id?

Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWorkMBean.java:44]: added

**Comment by ayushtkn** [comment]
**First Snap:**
image
**Second Snap**
image

**Third Snap**
image

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWork.java [hunk 5]
```diff
@@ -347,6 +348,27 @@ public String getReplStats() {
...
+      return "Got Error" + e.getMessage();
+    }
+  }
+
+  @Override
+  public Long getLastEventId() {
+    if (incrementalLoadTasksBuilder != null) {
+      return incrementalLoadTasksBuilder.eventTo();
+    }
+    return -1L;
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadWork.java:369]: this will be eventFrom for bootstrap

**Review by aasha — APPROVED** [review]

**Pull request merged**